dataframely/columns/array.py

```diff
@@ -0,0 +1,128 @@
+            metadata: A dictionary of metadata to attach to the column.
+        """
+        if primary_key:
+            raise ValueError(
+                "`primary_key=True` is not yet supported for the Array type."
+            )
+
+        if inner.primary_key or (
+            isinstance(inner, Struct)
+            and any(col.primary_key for col in inner.inner.values())
+        ):
+            raise ValueError(
+                "`primary_key=True` is not yet supported for inner types of the Array type."
+            )
+
+        # We disallow validation rules on the inner type since Polars arrays currently don't support .eval(). Converting
+        # to a list and calling .list.eval() is possible, however, since the shape can have multiple axes, the recursive
+        # conversion could have significant performance impact. Hence, we simply disallow inner validation rules.
+        # Another option would be to allow validation rules only for sampling, but not enforce them.
+        if inner.validation_rules(pl.col(alias or "")):
+            raise ValueError(
+                "Validation rules on the inner type of Array are not yet supported."
+            )
+
+        super().__init__(
+            nullable=nullable,
+            primary_key=False,
+            check=check,
+            alias=alias,
+            metadata=metadata,
+        )
+        self.inner = inner
+        self.shape = self._ensure_tuple(shape)
+
+    @property
+    def dtype(self) -> pl.DataType:
+        return pl.Array(self.inner.dtype, self.shape)
+
+    @staticmethod
+    def _ensure_tuple(value: int | tuple[int, ...]) -> tuple[int, ...]:
+        if isinstance(value, tuple):
+            return value
+        return (value,)
+
+    def sqlalchemy_dtype(self, dialect: sa.Dialect) -> sa_TypeEngine:
+        # NOTE: We might want to add support for PostgreSQL's ARRAY type or use JSON in the future.
+        raise NotImplementedError("SQL column cannot have 'Array' type.")
+
+    def _pyarrow_dtype_of_shape(self, shape: Sequence[int]) -> pa.DataType:
+        if shape:
+            size, *rest = shape
+            return pa.list_(self._pyarrow_dtype_of_shape(rest), size)
+        else:
+            return self.inner.pyarrow_dtype
+
+    @property
+    def pyarrow_dtype(self) -> pa.DataType:
+        return self._pyarrow_dtype_of_shape(self.shape)
+
+    def _sample_unchecked(self, generator: Generator, n: int) -> pl.Series:
+        # Sample the inner elements in a flat series
+        n_elements = n * cast(int, pl.Series(self.shape).product())
+        all_elements = self.inner.sample(generator, n_elements)
+
+        # Turn the "long" series into a series with arrays
+        it = iter(all_elements)
+
+        def _reshape_recursive(shape: Sequence[int]) -> Any:
+            if not shape:
+                return next(it)
+            rest: list[int]
+            n, *rest = shape
+            return [_reshape_recursive(rest) for _ in range(n)]
```

Comment:
I think this part here can be easier with native polars functionality (see https://docs.pola.rs/api/python/stable/reference/series/api/polars.Series.reshape.html) along the lines of:

```python
all_elements.reshape((n,) + self.shape)